src/security/Session.cc

```diff
@@ -180,13 +180,14 @@ CreateSession(const Security::ContextPointer &ctx, const Comm::ConnectionPointer
 }
 
 bool
-Security::CreateClientSession(const Security::ContextPointer &ctx, const Comm::ConnectionPointer &c, const char *squidCtx)
+Security::CreateClientSession(FuturePeerContext &ctx, const Comm::ConnectionPointer &c, const char *squidCtx)
 {
-    if (!c || !c->getPeer())
-        return CreateSession(ctx, c, Security::ProxyOutgoingConfig, Security::Io::BIO_TO_SERVER, squidCtx);
-
-    auto *peer = c->getPeer();
-    return CreateSession(ctx, c, peer->secure, Security::Io::BIO_TO_SERVER, squidCtx);
+    // TODO: We cannot make ctx constant because CreateSession() takes
+    // non-constant ctx.options (PeerOptions). It does that because GnuTLS
+    // needs to call PeerOptions::updateSessionOptions(), which is not constant
+    // because it compiles options (just in case) every time. To achieve
+    // const-correctness, we should compile PeerOptions once, not every time.
+    return CreateSession(ctx.raw, c, ctx.options, Security::Io::BIO_TO_SERVER, squidCtx);
```

Comment:
Q3: Should we postpone fixing old (but now marked) problems with PeerOptoons being modified at transaction handling time?

The alternative is to start revamping that extremely messy class so that we can make sure that the options are parsed before first use by transactions (and, hence, remain constant once they are parsed). This refactoring will require at least one dedicated PR. I am especially concerned with the affected experimental GnuTLS code that would be difficult to test and collaborate on. On the bright side, fixing those old problems would allow us to improve const-correctness of new PeerContext code now, when that code is added.

Reply:
Agreed on the postpone. At least it is out-of-scope for this PR.

FTR; the GnuTLS code has proven to be quite stable and is in use by a large number (~2k) of Debian installs. IMO we can stop calling the forward-proxy parts experimental as of v7/master.